libtransact/src/state/btree.rs

```diff
@@ -0,0 +1,181 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -----------------------------------------------------------------------------
+ */
+
+use crate::database::error::DatabaseError;
+use crate::database::lmdb::{Database, DatabaseReader, DatabaseWriter, LmdbDatabaseReaderCursor};
+use std::collections::BTreeMap;
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+
+#[derive(Clone)]
+pub struct BTreeBackEnd {
+    btree: Arc<Mutex<BTreeDatabase>>,
+}
+
+impl BTreeBackEnd {
+    pub fn new(indexes: &[&str]) -> BTreeBackEnd {
+        BTreeBackEnd {
+            btree: Arc::new(Mutex::new(BTreeDatabase::new(indexes))),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct BTreeDatabase {
+    pub main: BTreeMap<Vec<u8>, Vec<u8>>,
+    indexes: HashMap<String, BTreeMap<Vec<u8>, Vec<u8>>>,
+}
+
+impl BTreeDatabase {
+    pub fn new(indexes: &[&str]) -> BTreeDatabase {
+        let mut index_dbs = HashMap::with_capacity(indexes.len());
+        for name in indexes {
+            index_dbs.insert(String::from(name.as_ref()), BTreeMap::new());
+        }
+        BTreeDatabase {
+            main: BTreeMap::new(),
+            indexes: index_dbs,
+        }
+    }
+}
+
+impl Database for BTreeBackEnd {
+    fn get_reader(&self) -> Result<Box<dyn DatabaseReader>, DatabaseError> {
+        Ok(Box::new(BTreeReader {
+            db: self.btree.clone(),
+        }))
+    }
+    fn get_writer(&self) -> Result<Box<dyn DatabaseWriter>, DatabaseError> {
+        Ok(Box::new(BTreeWriter {
+            db: self.btree.clone(),
+        }))
+    }
+}
+
+pub struct BTreeReader {
+    db: Arc<Mutex<BTreeDatabase>>,
+}
+
+impl DatabaseReader for BTreeReader {
+    fn get(&self, key: &[u8]) -> Option<Vec<u8>> {
+        match self.db.lock().expect("error").main.get(key) {
+            Some(value) => Some(value.to_vec()),
+            None => None,
+        }
+    }
+
+    fn index_get(&self, index: &str, key: &[u8]) -> Result<Option<Vec<u8>>, DatabaseError> {
+        match self
+            .db
+            .lock()
+            .expect("error")
+            .indexes
+            .get(index)
+            .ok_or_else(|| DatabaseError::ReaderError(format!("Not an index: {}", index)))?
+            .get(key)
+        {
+            Some(value) => Ok(Some(value.to_vec())),
+            None => Ok(None),
+        }
+    }
+
+    /// Returns a cursor against the main database. The cursor iterates over
+    /// the entries in the natural key order.
+    fn cursor(&self) -> Result<LmdbDatabaseReaderCursor, DatabaseError> {
```

Comment:
I'm not sure yet how to implement `cursor` and `index_cursor` for the memory-backed state.
In the DatabaseReader Trait definition I cannot change it to use a generic type, such as <I: Iterator>, because that is not compatible with using DatabaseReader as an object. 
```
error[E0038]: the trait `database::lmdb::DatabaseReader` cannot be made into an object
   --> libtransact/src/database/lmdb.rs:129:5
    |
129 |     fn get_reader(&self) -> Result<Box<dyn DatabaseReader>, DatabaseError> {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `database::lmdb::DatabaseReader` cannot be made into an object
    |
    = note: method `cursor` has generic type parameters
    = note: method `index_cursor` has generic type parameters

``` 

I tried to change cursor and index_cursor to return  `Result<Box<dyn Iterator<Item = (Vec<u8>, Vec<u8>)>>, DatabaseError>` ... But I run into problem with the implementation of `LmdbDatabaseReaderCursor` in lmdb as it requires lifetime parameters, and we get conflicting lifetimes. 
```
error[E0495]: cannot infer an appropriate lifetime for lifetime parameter in function call due to conflicting requirements
   --> libtransact/src/database/lmdb.rs:215:14
    |
215 |             .cursor(self.db.main.clone())
    |              ^^^^^^
    |
note: first, the lifetime cannot outlive the anonymous lifetime #1 defined on the method body at 212:5...
   --> libtransact/src/database/lmdb.rs:212:5
    |
212 | /     fn cursor(&self) -> Result<Box<dyn Iterator<Item = (Vec<u8>, Vec<u8>)>>, DatabaseError> {
213 | |         let cursor = self
214 | |             .txn
215 | |             .cursor(self.db.main.clone())
...   |
218 | |         Ok(Box::new(LmdbDatabaseReaderCursor { access, cursor }))
219 | |     }
    | |_____^
note: ...so that reference does not outlive borrowed content
   --> libtransact/src/database/lmdb.rs:213:22
    |
213 |           let cursor = self
    |  ______________________^
214 | |             .txn
    | |________________^
    = note: but, the lifetime must be valid for the static lifetime...
    = note: ...so that the expression is assignable:
            expected std::result::Result<std::boxed::Box<(dyn std::iter::Iterator<Item=(std::vec::Vec<u8>, std::vec::Vec<u8>)> + 'static)>, _>
               found std::result::Result<std::boxed::Box<dyn std::iter::Iterator<Item=(std::vec::Vec<u8>, std::vec::Vec<u8>)>>, _>
```

`cursor` and `index_cursor` are not currently used in merkle... so a simple solution is to remove remove those functions from the DatabaseReader trait definition? Any other suggestions?